### PR TITLE
feat: process table

### PR DESCRIPTION
**Pull request by NhanChau2409**
### Linked issue

Closes [ducanhpham0312/zeevision-private#107
](https://github.com/ducanhpham0312/zeevision-private/issues/107)Closes ducanhpham0312/zeevision-private#36
### Description

- Basic struct of process table with GORM tags in [backend/internal/storage/table.go](https://github.com/ducanhpham0312/zeevision/pull/209/files#diff-50f637e0ef9e81a34d96de12629db4ce8afc287ec251282d35785eccac24cd14)

- Database's environment variables in [backend/internal/environment/environment.go](https://github.com/ducanhpham0312/zeevision/compare/%23107-process-table?expand=1#diff-258202835965b2305b07442c5028cf37f8f55ab1dbe3a822229652dae5a13c25)

- ConnectDb & CreateProcessTable functions in [backend/internal/storage/database.go](https://github.com/ducanhpham0312/zeevision/compare/%23107-process-table?expand=1#diff-5c46ec2ee947120f186c8301a3e07dd519213c5d89feb66fb4eea065e0733696)

- Connect & create table in [backend/cmd/zeevision/main.go](https://github.com/ducanhpham0312/zeevision/compare/%23107-process-table?expand=1#diff-ef6156588aa951dcf7867b9c147add8451ce17a243c5794d419e961fcf532abf)

- Unit test for connecting & creating table in [backend/internal/storage/storage_test.go](https://github.com/ducanhpham0312/zeevision/compare/%23107-process-table?expand=1#diff-ebcad93dd7a22af5938c7c6ff8c3122478c1567462ea4aaa1a9d4d959aafbff5) (which does not include "connecting db"-ed function)

### Testing

- Run `zeebee-test-bench` docker compose
- Run `docker-copmose up -d `
- Access `pgadmin`, instruction in backend README 
- On the right menu dropbox, choose `Servers/zeevision_db/Databases/zeevision_db/Shemas/public/Tables` 

If there is processes table that has columns as gorm struct 6 columns, this branch is well working

### CODE REVIEW

**Comment by NhanChau2409** [comment]
- Are any comments on the way I structuring DSN? I am not prettier sure is it clean or not. 

- I also fail backend/lints with `Error: G101: Potential hardcoded credentials (gosec)` as I search cause the database `user` & `password` are coded in `environment.go `which should instead store in `env` file or something.

File: backend/internal/environment/environment.go [hunk 1]
```diff
@@ -32,6 +32,18 @@ const (
...
 	EnvVarAPIAllowedOrigins = "ZEEVISION_API_ALLOWED_ORIGINS"
+	// Environment variable used to configure the database username
+	EnvVarDatabaseUser = "ZEEVISION_DATABASE_USER" //nolint:all
+	// Environment variable used to configure the database user password
+	EnvVarDatabasePassword = "ZEEVISION_DATABASE_PASSWORD" //nolint:all
+	// Environment variable used to configure the database name
+	EnvVarDatabaseName = "ZEEVISION_DATABASE_NAME"
+	// Environment variable used to configure if the database should be
+	// hosted.
+	EnvVarHostDatabase = "ZEEVISION_HOST_DATABASE"
```
Comment (kauppie) [inline backend/internal/environment/environment.go:43]: Is this comment description accurate? I think this was the host name or something, reading the README.

File: backend/internal/environment/environment.go [hunk 2]
```diff
@@ -32,6 +32,18 @@ const (
...
+	// Environment variable used to configure the port to use for the
+	// database.
+	EnvVarDatabasePort = "ZEEVISION_DATABASE_PORT"
```
Comment (kauppie) [inline backend/internal/environment/environment.go:46]: I think the comment here should something like that this port is used to connect to the database, not for hosting it like it currently sounds like.

File: backend/internal/environment/environment.go [hunk 3]
```diff
@@ -70,9 +88,18 @@ func init() {
...
 	setOrFallbackMap(EnvVarHostPlayground, DefaultHostPlayground, isOne)
 
-	setOrFallbackMap(EnvVarAPIAllowedOrigins, DefaultAPIAllowedOrigins, func(s string) ([]string, bool) {
-		return strings.Split(s, ","), true
-	})
+	setOrFallbackMap(
+		EnvVarAPIAllowedOrigins, DefaultAPIAllowedOrigins, func(s string) ([]string, bool) {
+			return strings.Split(s, ","), true
+		},
+	)
```
Comment (kauppie) [inline backend/internal/environment/environment.go:95]: Any reason to change this formatting? I think it was previously just fine.

File: backend/internal/environment/environment.go [hunk 4]
```diff
@@ -70,9 +88,18 @@ func init() {
...
+
+	setOrFallback(EnvVarDatabaseName, DefaultDatabaseName)
+
+	setOrFallbackMap(EnvVarDatabasePort, DefaultDatabasePort, parsePort)
+
+	setOrFallback(EnvVarHostDatabase, DefaultHostDatabase)
+
```
Comment (kauppie) [inline backend/internal/environment/environment.go:102]: You can leave out this last empty line.

File: backend/internal/environment/environment.go [hunk 5]
```diff
@@ -112,11 +139,38 @@ func APIAllowedOrigins() []string {
...
 // Helper to save environment variable value if it has been set.
 func setOrFallback(envVar string, fallback string) {
-	setOrFallbackMap(envVar, fallback, func(s string) (string, bool) {
-		return s, true
-	})
+	setOrFallbackMap(
+		envVar, fallback, func(s string) (string, bool) {
+			return s, true
+		},
+	)
```
Comment (kauppie) [inline backend/internal/environment/environment.go:173]: Is this change needed?

File: backend/internal/storage/database.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import (
+	"fmt"
+
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+)
+
+// Configuration to connect to database
+type DsnConfig struct {
+	// Username to login database
```
Comment (kauppie) [inline backend/internal/storage/database.go:12]: "Username to login database" -> "Username used to login to the database"

File: backend/internal/storage/database.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	User string
+	// User password to verify user
```
Comment (kauppie) [inline backend/internal/storage/database.go:14]: "User password to verify user" -> "Password of the logging in user"

File: backend/internal/storage/database.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+	Password string
+	// Database name to be accessed
+	DatabaseName string
+	// The host used to host database
+	Host string
+	// The port used to connect database
```
Comment (kauppie) [inline backend/internal/storage/database.go:20]: "The port used to connect database" -> "The port used to connect to the database"

File: backend/internal/storage/database.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// Create a valid DSN string to connect database from DsnConfig
+func (config *DsnConfig) NewDsn() string {
+	dsn := fmt.Sprintf(
+		"user=%s password=%s dbname=%s host=%s port=%d",
+		config.User, config.Password, config.DatabaseName, config.Host, config.Port,
+	)
+	return dsn
+}
```
Comment (kauppie) [inline backend/internal/storage/database.go:31]: Now that you made this into a method, I think better name for this could be just `String()` as it only converts it into a string form.

File: backend/internal/storage/database.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+// Connect to database by DsnConfig
+func ConnectDb(dsnConfig DsnConfig) (*gorm.DB, error) {
+
```
Comment (kauppie) [inline backend/internal/storage/database.go:35]: You can leave out this empty line.

File: backend/internal/storage/database.go [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+	dsn := dsnConfig.NewDsn()
+
+	db, err := gorm.Open(postgres.Open(dsn), &gorm.Config{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to database: %w", err)
+
```
Comment (kauppie) [inline backend/internal/storage/database.go:41]: You can leave out this empty line.

File: backend/internal/storage/database.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// Connect to database by DsnConfig
+func ConnectDb(dsnConfig DsnConfig) (*gorm.DB, error) {
+
+	dsn := dsnConfig.NewDsn()
+
+	db, err := gorm.Open(postgres.Open(dsn), &gorm.Config{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to database: %w", err)
```
Comment (kauppie) [inline backend/internal/storage/database.go:40]: Nice. This is the error formatting style you often see when reading Go code:
- lowercase
- some error message understandable from the context
- inner error appended to the end separated by a colon.

Nothing to change here.

File: backend/internal/storage/storage_test.go [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+const DSN = "user=user password=password dbname=zeevision_db host=postgres port=5432"
+
+func TestNewDsn(t *testing.T) {
+	dsnConfig := DsnConfig{
+		User:         "user",
+		Password:     "password",
+		DatabaseName: environment.DatabaseName(),
+		Host:         environment.HostDatabase(),
+		Port:         environment.DatabasePort(),
```
Comment (kauppie) [inline backend/internal/storage/storage_test.go:18]: You can use hard-coded strings here since this is a test and shouldn't depend on the context.

File: backend/internal/storage/storage_test.go
Comment (kauppie) [inline backend/internal/storage/storage_test.go:1]: This file should be called `database_test.go` since it's testing functionality defined in `database.go`.

File: backend/internal/storage/table.go [hunk 14]
```diff
@@ -0,0 +1,11 @@
+package storage
+
+import "gorm.io/gorm"
+
+// Process model struct for database table
+type Process struct {
+	gorm.Model
+	ProcessID    uint   `gorm:"unique;not null"`
+	ProcessKey   uint   `gorm:"unique;not null"`
```
Comment (kauppie) [inline backend/internal/storage/table.go:9]: The incoming types from Kafka will be `int64`, so these should be too to fit those values.

**Review by kauppie — CHANGES_REQUESTED** [review]
Seems to be working how you described it. But this breaks `zeevision` itself since it doesn't have retry functionality when connecting to the database. `zeevision` should retry the connection to the database until it succeeds.

Nice that you set most variables using environment variables for ease of configuration.

File: backend/cmd/zeevision/main.go [hunk 15]
```diff
@@ -6,10 +6,30 @@ import (
...
 	"github.com/ducanhpham0312/zeevision/backend/internal/environment"
+	"github.com/ducanhpham0312/zeevision/backend/internal/storage"
 )
 
 // Entry point for the application.
 func main() {
+
+	dsnConfig := storage.DsnConfig{
+		User:         "zeevision_user",
+		Password:     "zeevision_pass",
```
[the remark this answers is not in the document]
Comment (kauppie) [reply backend/cmd/zeevision/main.go:17]: Remember to change `docker-compose.yml` to set these environment variables!

**Review by ducanhpham0312 — COMMENTED** [review]
An extra comment: Please update the PR description to use Conventional Commit syntax. Should be something like `feat: process table` here.

File: backend/internal/storage/database.go (hunk 9 above)
Comment (kauppie) [reply backend/internal/storage/database.go:31]: That was kind of the idea since `DsnConfig` only job at moment is to be a typed interface for the DSN string.

File: backend/cmd/zeevision/main.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (NhanChau2409) [reply backend/cmd/zeevision/main.go:17]: Can you explain more specifically ?

[the remark this answers is not in the document]
Comment (kauppie) [reply backend/cmd/zeevision/main.go:17]: Set User and Password from environment variables instead of hard-coding them in. Don't give User nor Password any default values like other variables in environment.go have.

**Review by kauppie — CHANGES_REQUESTED** [review]
There is PR #222 which can be merged to fix this branch. After that everything seems fine to me.
